stew/results.nim

```diff
@@ -573,3 +573,32 @@ template `?`*[T, E](self: Result[T, E]): T =
   if v.isErr: return err(typeof(result), v.error)
 
   v.value
+
+# Options compatibility
+
+import options
+export options
+
+when defined(resultsExperimental):
+  {.pragma: resexp.}
+else:
+  {.pragma: resexp, deprecated: "experimental".}
+
+func ok*[T, E](r: Result[T, E]): Option[T] {.resexp.} =
```

Comment:
I find the naming of this somewhat confusing. I use `(var result).ok` to set the result to the OK state, but calling `ok()` gives me an `Option`?

Reply:
yeah I thought about that, though strictly it's `res.ok()` and `res.ok(value)` - it reads kind of nice when doing the `in` trick - I'm more sceptical about `(var res).ok(value)` to be honest - it's there because we don't have the nice enum constructors where you only need to know either T or E - suggestions?

Reply:
It's just `result.ok()` in the `void` case, but then `Option[void]` is probably pointless. 
Nevertheless, why don't you go with a more obvious name such as `getOpt`?

Reply:
I tend to use the following syntax in my own code:

```nim
return ok foo()
...
return err "bar"
```

Or the versions without `return` where that's appropriate, so indeed the `var` overloads may be less important.

Reply:
`getOpt` / `toOption` no because both `ok` and `err` return an option so it's a bit ambiguous - also, it fits `isOk`/`isErr` - ie everything is within the same "name" world.

the `var` overloads are there for when you're not returning, but using `Result` outside of a return value, really, and `typeof(result)` is meaningless or wrong.

also, `if X in res.toOption` doesn't look as nice.

Reply:
the void case, that's a good point. hm.

Reply:
`toOption` makes sense to me, because I don't have to guess which `ok` variant is executed there.

Reply:
`setOk`/`setErr` for `var`? or remove them indeed and readd when there's a good use case

Reply:
But when is the `toOption` construct useful? Why are we giving it more priority by assigning the shortest possible name?

Reply:
well, options and results are used in similar situations, so they complete each other in a way - it's often useful to move between the two in a friction-less way.

but the main reason for not liking the `toOption` name is that it biases the interface towards the `ok` case which a bit defeats they purpose of `Result`.